### ISSUE TITLE
best practices regarding code structure

### ISSUE THREAD

**Issue by Remo** [flat3/lodata#129]
I'm playing around with lodata again, everything works very well and a lot of the OData specification has been implemented which is great.

One thing I'm struggling with, is the whole code structure.

**Operations**

If I extend a class from `Operation` I basically build a global class, but most of the time an operation is specific to a model. If I have a model called Person, I would want the operation `greet` to be part of that model/class and not other model.

In pseudo code I'd probably want something like this:

```
class Person extends BaseRepository {
   model = PersonModel::class

   greet() {
      ...
   }
}
```

And then register the whole class, including the public methods which would turn into operations. That way I can nicely group my code which would make it easier for new people on my project to find the relevant pieces.

I'm aware that I can bind an entity, but this again looks global: https://lodata.io/modelling/operations.html#binding-entities

**Overriding standard behaviour**

This seems to be a use-case which I need a lot. If I take my pseudo code from above, I might want to turn it into something like this:

```
class Person extends BaseRepository {
   model = PersonModel::class

   greet() {
      ...
   }

   onSave() {
      fullName = first_name + middle_name + last_name
   }
}
```

---

In other words: 
* While I don't want to have classes with hundreds of lines, I also don't want to have a single file per method.
* I'd like to keep things together from a business object point of view.

Maybe I could create a magic object register method that scans a bunch of annotations and register the entity as well as certain operations, but I wonder what others are doing to structure their lodata code.

**Comment by 27pchrisl** [comment]
Hi @Remo,

Exposing an Eloquent method as a bound operation on an entity in the way you describe is 100% the way this should be implemented, however as you point out it's not there yet 😉.

During Lodata's "discovery" of Eloquent models, it should look for methods decorated with PHP attributes that indicate the methods should be exposed as operations, and just dynamically generate those operation classes on the model with the right parameters - do you think this would work?

I am definitely interested in hearing ideas of different code patterns for creating operations and other pieces of the model. Lodata is still a relatively young project and can be changed based on what works in the real world.

**Comment by Remo** [comment]
There's just one thing I'm wondering about. Exposing decorated eloquent models methods would be cool, but it would only work if they are actually there. What if I'm using a repository pattern that wraps my model? I have to admit that in most projects I'm not mocking my models, but if the project gets large, a repository pattern with mocks can help and then it would fail.

Having just said that, I think this is probably a special use-case we can't easily support in lodata. A repository class can have any kind of method, if there's no contract, lodata wouldn't even know how to fetch a single model record. I guess lodata could ship with an interface that allows us to build repositories, but that's maybe a bit too opinionated.

Right now I'd say let's go with an automatic discovery of decorated eloquent model methods and make it easy for others, to write their own discovery method in case they want to support a custom repository class.

**Comment by Remo** [comment]
What about a really lightweight interface that simply returns the model?

```php
<?php
use Illuminate\Database\Eloquent\Model;

interface ILodataEloquentRepository {
   getModel(): Model;
}
```

That way I could easily create a repository myself with custom methods and still make lodata happy by giving it a straight connection to the model:

```php
<?php
#[LodataRepository]
class Flight implements ILodataEloquentRepository {
   public function getModel(): Model {
     return FlightModel::class;
   }

   #[LodataOperation]
   public function land() {
      ...
   }
}
```

And for those who would want to use their models directly we could go with something like this:

```php
<?php
use Illuminate\Database\Eloquent\Model;

#[LodataModel]
class Flight extends Model {
}
```

**Comment by 27pchrisl** [comment]
In this pattern, what's the right way of calling `land()` so it gets access to the right instance of the model it needs to operate on?

**Comment by Remo** [comment]
I'd say: `flights(1)/land`

**Comment by 27pchrisl** [comment]
Ah I meant in PHP 😄  how should the instance of the model be passed to land?

**Comment by Remo** [comment]
Oh yes, that makes more sense 😄 

Maybe we could do something like this:

```php
<?php
use Illuminate\Database\Eloquent\Model;

interface ILodataEloquentRepository {
   setInstance(Model $instance);
   getInstance(): Model;
   getModel(): Model;
}
```

Otherwise we'd have to pass it as an argument:
```php
#[LodataOperation]
public function land(Model $instance) {
   ...
}
```

I'd probably prefer the second approach, but either way, it becomes a little bit more opinionated.
It might be better to focus on traditional eloquent models?

**Comment by 27pchrisl** [comment]
Hi @Remo,

I have tagged https://github.com/flat3/lodata/releases/tag/v4.0.0-beta.1 which rewrites how operations are structured.

There are samples here: https://github.com/flat3/lodata/tree/4.x/tests/Operations
and here: https://github.com/flat3/lodata/blob/4.x/tests/Models/Airport.php
and here: https://github.com/flat3/lodata/blob/4.x/tests/Models/Repository.php

This should show how you can tag your methods with attributes, and then pass either the class name or a class instance to `Lodata::discover($class)` as described in https://github.com/flat3/lodata/blob/4.x/tests/Unit/Operation/DiscoveryTest.php

Once the interface is solid I will update the documentation.

Any feedback appreciated!

**Comment by Remo** [comment]
Wow this is great! At first glance this is exactly what I was hoping to have.

I haven't tested it, but will this work with PHP 7.3? I've just checked https://github.com/flat3/lodata/blob/4.x/composer.json and saw that we still allow it.

Do you think there's a way to find classes by an attribute like `#[LodataClass]` ? From what I understand it would be a bit of a brute force scan. Being able to declare the classes with an attribute would feel more consistent, it's what I'm used from my .NET projects, but I guess with PHP we'd have to scan all declared classes and then cache it?

**Comment by 27pchrisl** [comment]
Yes you can certainly use the functionality with 7.3, you just can't use attributes to automatically configure for you. All the attributes are doing are describing how an `Operation` should be configured.

I added tests at https://github.com/flat3/lodata/blob/4.x/tests/Unit/Operation/Repository73Test.php that test you can still create these operations in 7.3.

For finding classes decorated with an attribute you could do something as you're describing but I'm unsure what the performance would be. You could also only do it over all classes that had already been loaded, so if you had a class that hadn't yet been autoloaded yet you couldn't discover it, so you'd still have to add it for discovery manually.

**Comment by Remo** [comment]
>  Yes you can certainly use the functionality with 7.3, you just can't use attributes to automatically configure for you. All the attributes are doing are describing how an Operation should be configured.

That's great!

> For finding classes decorated with an attribute you could do something as you're describing but I'm unsure what the performance would be. You could also only do it over all classes that had already been loaded, so if you had a class that hadn't yet been autoloaded yet you couldn't discover it, so you'd still have to add it for discovery manually.

I guess not being able to find such classes easily is a restriction we have with PHP. No big deal. I'll try to add some attributes to my classes in my lodata test project, but I'll need a bit of time to get that done.

**Comment by Remo** [comment]
I've upgraded a demo project of mine to the latest beta of lodata. The main functionality was up and running in no time, but one thing I'm not sure about.

I previously got this in my code:
```
\Lodata::discoverEloquentModel(\App\Models\Project::class)->discoverRelationship('issues');
```

I replaced it with `Lodata::discover` which works fine for project model data, but the relationship is missing.

I wonder, shouldn't we have something like `#[LodataRelationship]`? Or what did you have in mind regarding classes with relationships?

```php
<?php

namespace Flat3\Lodata\Tests\Models;

use Flat3\Lodata\Attributes\LodataFunction;
use Illuminate\Database\Eloquent\Model;

/**
 * @property string $name
 * @property string $code
 */
class Airport extends Model
{
    public $timestamps = false;
    protected $guarded = [];
    protected $casts = [
        'construction_date' => 'date',
        'sam_datetime' => 'datetime',
        'open_time' => 'timestamp',
        'review_score' => 'float',
        'is_big' => 'boolean',
    ];

    public function country()
    {
        return $this->belongsTo(Country::class);
    }

    #[LodataRelationship]
    public function flights()
    {
        return $this->hasMany(Flight::class, 'origin', 'code');
    }

    public function scopeModern($query)
    {
        return $query->where('construction_date', '>', '1940-01-01');
    }

    #[LodataFunction]
    public function op1(): ?string
    {
        return $this->name;
    }

    #[LodataFunction]
    public function op2(string $prefix): string
    {
        return $prefix.$this->code;
    }
}
```

**Comment by 27pchrisl** [comment]
You're absolutely right, we should have that! Shouldn't be hard to add...

**Comment by 27pchrisl** [comment]
Okay that's all added, I tagged the https://github.com/flat3/lodata/releases/tag/v4.0.0 release that includes this.

**Comment by Remo** [comment]
I've tested it myself, it works perfect! What a great step forward.

I don't think there's anything right now that I would change. This issue can be closed from my point of view.

**Comment by 27pchrisl** [comment]
Thanks @Remo I appreciate the feedback!

**Comment by bimix** [comment]
this is not working for me. I added the #[LodataRelationship] but still the results only include the model data, not the relation data.
